File: sdk/integrations/django/middleware.py
"""
Django Middleware for Semantys Cache

Automatically caches OpenAI API calls in Django applications.

Requires the main Semantys SDK: pip install semantys
"""
import json
import os
import time
from typing import Optional, Callable

from django.http import HttpRequest, HttpResponse, JsonResponse
from django.utils.deprecation import MiddlewareMixin


class SemanticCacheMiddleware(MiddlewareMixin):
    """
    Django middleware for automatic semantic caching.

    Add to MIDDLEWARE in settings.py::

        MIDDLEWARE = [
            'semantys.integrations.django.SemanticCacheMiddleware',
            ...
        ]

    Configuration (settings.py)::

        SEMANTYS_CACHE_API_KEY = 'sc-your-key'
        SEMANTYS_CACHE_BASE_URL = 'https://api.semantys.ai'  # optional
        SEMANTYS_CACHE_PATHS = ['/v1/chat/completions']        # optional
    """

    def __init__(self, get_response: Callable):
        super().__init__(get_response)
        self.get_response = get_response

        from django.conf import settings
        self.api_key = (
            getattr(settings, 'SEMANTYS_CACHE_API_KEY', None)
            or os.getenv('SEMANTYS_API_KEY')
        )
        self.base_url = (
            getattr(settings, 'SEMANTYS_CACHE_BASE_URL', None)
            or os.getenv('SEMANTYS_API_URL', 'https://api.semantys.ai')
        )
        self.cache_paths = getattr(
            settings, 'SEMANTYS_CACHE_PATHS', ['/v1/chat/completions']
        )

        if not self.api_key:
            raise ValueError(
                "API key required. Set SEMANTYS_CACHE_API_KEY in settings "
                "or SEMANTYS_API_KEY env var."
            )

        try:
            from semantys import SemantysCache
            self._cache = SemantysCache(api_key=self.api_key, base_url=self.base_url)
        except ImportError:
            raise ImportError(
                "semantys package is required. Install with: pip install semantys"
            )

    def process_request(self, request: HttpRequest) -> Optional[HttpResponse]:
        if request.path not in self.cache_paths or request.method != 'POST':
            return None

        try:
            body = json.loads(request.body)
            messages = body.get('messages', [])
            user_messages = [m for m in messages if m.get('role') == 'user']
            if not user_messages:
                return None

            prompt = user_messages[-1].get('content', '')
            if not prompt:
                return None

            resp = self._cache.chat.completions.create(
                model=body.get('model', 'gpt-4o-mini'),
                messages=messages,
                temperature=body.get('temperature', 0.2),
            )
            if resp.meta and resp.meta.hit in ('exact', 'semantic'):
                return JsonResponse({
                    'id': resp.id,
                    'object': resp.object,
                    'created': resp.created,
                    'model': resp.model,
                    'choices': [
                        {
                            'index': c.index,
                            'message': {'role': c.message.role, 'content': c.message.content},
                            'finish_reason': c.finish_reason,
                        }
                        for c in resp.choices
                    ],
                    'usage': {
                        'prompt_tokens': resp.usage.prompt_tokens if resp.usage else 0,
                        'completion_tokens': resp.usage.completion_tokens if resp.usage else 0,
                        'total_tokens': resp.usage.total_tokens if resp.usage else 0,
                    },
                    'meta': {
                        'hit': resp.meta.hit,
                        'similarity': resp.meta.similarity,
                        'latency_ms': resp.meta.latency_ms,
                    },
                })
        except Exception:
            pass

        return None

File: sdk/integrations/express/middleware.js
/**
 * Express.js Middleware for Semantis Cache
 * 
 * Automatically caches OpenAI API calls in Express applications.
 */

const { SemanticCache } = require('semantis-cache');

/**
 * Create semantic cache middleware for Express
 * 
 * @param {Object} options - Middleware options
 * @param {string} options.apiKey - Semantis AI API key
 * @param {string} [options.baseUrl] - API base URL (default: https://api.semantis.ai)
 * @param {string[]} [options.cachePaths] - Paths to cache (default: ["/v1/chat/completions"])
 * @returns {Function} Express middleware function
 * 
 * @example
 * const express = require('express');
 * const semanticCacheMiddleware = require('semantis-cache/integrations/express');
 * 
 * const app = express();
 * app.use(express.json());
 * 
 * // Add middleware
 * app.use(semanticCacheMiddleware({
 *   apiKey: 'sc-your-key',
 *   cachePaths: ['/v1/chat/completions']
 * }));
 */
function semanticCacheMiddleware(options) {
  const {
    apiKey,
    baseUrl = process.env.SEMANTIS_API_URL || 'https://api.semantis.ai',
    cachePaths = ['/v1/chat/completions'],
  } = options;

  if (!apiKey) {
    throw new Error('API key is required. Provide it in options or set SEMANTIS_API_KEY environment variable.');
  }

  // Initialize cache client
  const cache = new SemanticCache({
    apiKey: apiKey || process.env.SEMANTIS_API_KEY,
    baseUrl: baseUrl,
  });

  return async (req, res, next) => {
    // Check if this path should be cached
    if (!cachePaths.includes(req.path)) {
      return next();
    }

    // Only cache POST requests (OpenAI API calls)
    if (req.method !== 'POST') {
      return next();
    }

    try {
      // Extract prompt from request body
      const messages = req.body.messages || [];
      if (messages.length === 0) {
        return next();
      }

      // Get the last user message
      const userMessages = messages.filter(msg => msg.role === 'user');
      if (userMessages.length === 0) {
        return next();
      }

      const prompt = userMessages[userMessages.length - 1].content;
      if (!prompt) {
        return next();
      }

      // Check cache
      try {
        const response = await cache.query(prompt, req.body.model || 'gpt-4o-mini');

        // If cache hit, return cached response
        if (response.cacheHit === 'exact' || response.cacheHit === 'semantic') {
          // Format as OpenAI response
          const cachedResponse = {
            id: `chatcmpl-cached-${hashCode(prompt)}`,
            object: 'chat.completion',
            created: Math.floor(Date.now() / 1000),
            model: req.body.model || 'gpt-4o-mini',
            choices: [{
              index: 0,
              message: {
                role: 'assistant',
                content: response.answer,
              },
              finish_reason: 'stop',
            }],
            usage: {
              prompt_tokens: 0,
              completion_tokens: 0,
              total_tokens: 0,
            },
            meta: {
              hit: response.cacheHit,
              similarity: response.similarity,
              latency_ms: response.latencyMs,
            },
          };

          return res.json(cachedResponse);
        }
      } catch (error) {
        // If cache fails, continue to original endpoint
        console.warn('Cache error:', error.message);
      }
    } catch (error) {
      // If anything fails, continue to original endpoint
      console.warn('Middleware error:', error.message);
    }

    // If no cache hit, continue to original endpoint
    next();
  };
}

/**
 * Simple hash function for prompt
 */
function hashCode(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash; // Convert to 32-bit integer
  }
  return Math.abs(hash).toString(36);
}

module.exports = semanticCacheMiddleware;

